packages/jest-mock/src/index.ts

```diff
@@ -1008,40 +1008,48 @@ export class ModuleMocker {
     return fn;
   }
 
-  spyOn<T extends object, M extends PropertyLikeKeys<T>>(
+  spyOn<
+    T extends object,
+    K extends PropertyLikeKeys<Required<T>>,
+    V extends Required<T>[K],
+  >(object: T, methodKey: K, accessType: 'get'): SpyInstance<() => V>;
+
+  spyOn<
+    T extends object,
+    K extends PropertyLikeKeys<Required<T>>,
+    V extends Required<T>[K],
+  >(object: T, methodKey: K, accessType: 'set'): SpyInstance<(arg: V) => void>;
+
+  spyOn<
+    T extends object,
+    K extends ConstructorLikeKeys<Required<T>>,
+    V extends Required<T>[K],
+  >(
     object: T,
-    methodName: M,
-    accessType: 'get',
-  ): SpyInstance<() => T[M]>;
-
-  spyOn<T extends object, M extends PropertyLikeKeys<T>>(
-    object: T,
-    methodName: M,
-    accessType: 'set',
-  ): SpyInstance<(arg: T[M]) => void>;
-
-  spyOn<T extends object, M extends ConstructorLikeKeys<T>>(
-    object: T,
-    methodName: M,
-  ): T[M] extends ClassLike
-    ? SpyInstance<(...args: ConstructorParameters<T[M]>) => InstanceType<T[M]>>
+    methodKey: K,
+  ): V extends ClassLike
+    ? SpyInstance<(...args: ConstructorParameters<V>) => InstanceType<V>>
     : never;
 
-  spyOn<T extends object, M extends MethodLikeKeys<T>>(
+  spyOn<
+    T extends object,
+    K extends MethodLikeKeys<Required<T>>,
+    V extends Required<T>[K],
```

Comment:
Lines with `Required` do all the magic, because they turn types like `(() => void)) | undefined` into `() => void)`. Something what wasn’t a function, becomes one. This part was not working, otherwise logic is the same.